### PR TITLE
[RLlib] Enable counting of SampleBatch length by traversing nested structures

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
Signed-off-by: Artur Niederfahrenhorst <artur@anyscale.com>

## Why are these changes needed?

TLDR: Today we count only by checking the length of top-level values of the SampleBatch. If the value is tuple or a dict, we skip.
This may result in a .count of 0 if there are only nested values. For example in a batch that only carries an OBS with a nested structure. This leads to issues further down the line since the count of the sample batch is not zero, but whatever is the length of arrays within the nested OBS.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/sample_batch.py [hunk 1]
```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
...
+                else:
+                    # If we were able to figure out a length, all lengths of
+                    # elements of nested structure must be the same
+                    assert all(len(sub_space) == len(v[0]) for sub_space in v)
+                continue
+
             # TODO: Drop support for lists as values.
             # Convert lists of int|float into numpy arrays make sure all data
             # has same length.
             if isinstance(v, list):
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:168]: make this elif?

File: rllib/policy/sample_batch.py [hunk 2]
```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
...
+                if k == SampleBatch.INFOS:
+                    # Don't attempt to count on infos since we make no assumptions
+                    # about its content
+                    continue
+                v = tree.flatten(v)
+                try:
+                    # Add one of the elements length, since they are all the same
+                    len_ = len(v[0])
+                    if len_:
+                        lengths.append(len_)
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:156]: remove if here:
```
lengths.append(len(v[0]))
```

File: rllib/policy/sample_batch.py [hunk 3]
```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
...
 
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            if isinstance(v, (dict, tuple)):
+                if k == SampleBatch.INFOS:
+                    # Don't attempt to count on infos since we make no assumptions
+                    # about its content
+                    continue
+                v = tree.flatten(v)
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:151]: call it v_flat to not confuse it with the original v?

File: rllib/policy/sample_batch.py [hunk 4]
```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
...
             # Convert lists of int|float into numpy arrays make sure all data
             # has same length.
             if isinstance(v, list):
                 self[k] = np.array(v)
 
-            # Try to infer the "length" of the SampleBatch by finding the first
-            # value that is actually a ndarray/tensor. This would fail if
-            # all values are nested dicts/tuples of more complex underlying
-            # structures.
             try:
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:171]: You have redundancy of code here. Can you re-write this in a way you don't repeat yourself?

File: rllib/policy/sample_batch.py [hunk 5]
```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
...
+                    len_ = len(v[0])
+                    if len_:
+                        lengths.append(len_)
+                except Exception:
+                    pass
+                else:
+                    # If we were able to figure out a length, all lengths of
+                    # elements of nested structure must be the same
+                    assert all(len(sub_space) == len(v[0]) for sub_space in v)
+                continue
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:163]: Also you won't need continue if the code is re-structured a bit. (with elif and removal of the redundancy)

File: rllib/policy/tests/test_sample_batch.py [hunk 6]
```diff
@@ -424,6 +424,67 @@ def get_mismatched_types(v):
...
+            ),
+            (
+                {
+                    SampleBatch.INFOS: {
+                        "a": np.array([[1], [2], [3]]),
+                        "b": np.array([[0], [0], [1]]),
+                        "c": np.array([[4], [5], [6]]),
+                    }
+                },
+                0,  # This should have a length of zero, since we can not determine it
```
Comment (kouroshHakha) [inline rllib/policy/tests/test_sample_batch.py:459]: nit: since we ignore INFO not since we cannot determine it.

File: rllib/policy/tests/test_sample_batch.py [hunk 7]
```diff
@@ -424,6 +424,67 @@ def get_mismatched_types(v):
             # should raise an error if framework is not torch
             s.to_device(cuda_if_possible, framework="tf")
 
+    def test_count(self):
```
Comment (kouroshHakha) [inline rllib/policy/tests/test_sample_batch.py:427]: nice test.

**Review by kouroshHakha — APPROVED** [review]
Thanks, needs a little bit of code refactor and is approved subject to tests passing.

File: rllib/policy/sample_batch.py (hunk 4 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:171]: I don't think that's possible because of the `else:` after the try/except.

File: rllib/policy/sample_batch.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:156]: This would change the logic of the whole function because we further down depend on lengths[0].
Right now, `lengths` can never contain zeros.

Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:156]: Don't get me wrong - I want this and did it already before submitting.
But this would enlargen the scope of what can happen here significantly and even break user code.

File: rllib/policy/sample_batch.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:163]: Conditioning this on my above comments.

File: rllib/policy/sample_batch.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/policy/sample_batch.py]: oh so len_ should not be appended if it's zero. I see. makes sense.

File: rllib/policy/sample_batch.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/policy/sample_batch.py:170]: How about this:
```
if k == SampleBatch.INFOS: continue

if isinstance(v, list):
    self[k] = np.array(v)
else:
    v_flat = tree.flatten(v) if isinstance(v, (dict, tuple)) else [v]
    try:
        len_ = len(v_flat[0])
        if len_:
            lengths.append(len_)
    except Exception:
        pass
    else:
        assert all(len(sub_space) == len(v_flat[0]) for sub_space in v_flat)
```

Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:170]: I think the else
```
if isinstance(v, list):
    self[k] = np.array(v)
else: <<<---
```
does not make sense in this case but I get the idea, thanks!

Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:170]: Also PyCharm just reminded me that `if k == SampleBatch.INFOS: continue` is not PEP conform

File: rllib/policy/sample_batch.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:163]: Because of the _len thing above, I'll leave it in!

Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:163]: Removed! Thanks

File: rllib/policy/sample_batch.py [hunk 8]
```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
...
+                len_ = len(v_list[0])
                 if len_:
                     lengths.append(len_)
             except Exception:
                 pass
+            else:
+                # If we were able to figure out a length, all lengths of
+                # elements of nested structure must be the same
+                try:
+                    same_lengths = all(
```
Comment (ArturNiederfahrenhorst) [inline rllib/policy/sample_batch.py:175]: This works if all nested elements are arrays (which is what we want).

File: rllib/policy/tests/test_sample_batch.py [hunk 9]
```diff
@@ -372,14 +372,15 @@ def test_to_device(self):
         s = SampleBatch(
             {
                 "a": np.array([1, 2]),
-                "b": {"c": (np.array([4, 5]), np.array([5, 6])), "e": [{"f": None}]},
```
Comment (ArturNiederfahrenhorst) [inline rllib/policy/tests/test_sample_batch.py:375]: @sven1977 I could not think of a use case for a None here and neither could Kourosh. Please ack.

Comment (ArturNiederfahrenhorst) [reply rllib/policy/tests/test_sample_batch.py:375]: Allowing Nones makes counting logic more convoluted and I believe it should simply not be supported.
We want to enforce tensors here in the long run. Wonky stuff can go into INFOS.

File: rllib/policy/sample_batch.py [hunk 10]
```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
...
+                        len(sub_space) == len(v_list[0]) for sub_space in v_list
+                    )
+                except TypeError:
+                    # If input contains scalar arrays (that don't have a length),
+                    # they should all be scalar arrays
+                    same_lengths = all([sub_space.size == 0 for sub_space in v_list])
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:181]: we are gonna take a look at this case together tomorrow.

**Review by kouroshHakha — APPROVED** [review]

File: rllib/policy/tests/test_sample_batch.py (hunk 9 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/tests/test_sample_batch.py:375]: (Not allowing Nones here does not break our tests)

File: rllib/policy/sample_batch.py (hunk 10 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:181]: Awesome! I hope we can merge this and move forward with connectors also.
The PR that highlights the wonky tests is linked now.

File: rllib/policy/sample_batch.py [hunk 11]
```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
...
-            # structures.
+            if k == SampleBatch.INFOS:
+                # Don't attempt to count on infos since we make no assumptions
+                # about its content
+                continue
+
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            v_list = tree.flatten(v) if isinstance(v, (dict, tuple)) else [v]
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:158]: maybe you don't need this if ... else. tree.flatten specifically converts single obj into [obj]
```
>>> tree.flatten(np.array([3, 4]))
[array([3, 4])]
```

File: rllib/policy/sample_batch.py (hunk 10 above)
Comment (gjoliver) [reply rllib/policy/sample_batch.py:181]: stupid question, what is a scalar array??

File: rllib/policy/sample_batch.py [hunk 12]
```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
...
+            # TODO: Drop support for lists and Numbers as values.
+            # If v_list contains lists or Numbers, convert them to arrays, too.
+            v_list = [
+                np.array(_v) if isinstance(_v, (Number, list)) else _v for _v in v_list
+            ]
             try:
-                len_ = len(v) if not isinstance(v, (dict, tuple)) else None
+                # Add one of the elements' length, since they are all the same
+                len_ = len(v_list[0])
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:166]: use ``_len``? underscore suffix is not part of the style guide,

File: rllib/policy/sample_batch.py [hunk 13]
```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
...
                 if len_:
                     lengths.append(len_)
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:168]: just to make sure I understand this, we queue this ``len_`` only if ``len(xxx)`` does not throw error.
so it's possible to mix up things that have lengths and not have lengths?
asking because I see that you want to get rid of the ``None`` support below.

**Review by gjoliver — COMMENTED** [review]
looks pretty reasonable.
minor comment at a high level, I would encourage us to create a static util function for this.
so we can reduce the length of the constructor, also unit test the util function instead of the whole SampleBatch.

File: rllib/policy/sample_batch.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:164]: I have one concern here: What if we have a sample batch that carries RNN states?
In those batches, the e.g. OBS usually have structure `[BxT, h]` (h = obs dim), whereas the "state_in_0" key has a structure of only `[B, h]` (or even more complicated: `[BxT/[max-seq-len], h]`). If this new counting procedure now finds one of the state fields first and use the len of the data for its own "count", then we get the count wrong.
Rough fix would be to skip `state_(in|out)_\d+` keys, just like we do with INFOS.

Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:164]: @sven1977 That would certainly make it more clean.
On master, we always use the first found length.
I'm guessing that this only worked because typically people put observations first.

File: rllib/policy/sample_batch.py (hunk 13 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:168]: Scalar np arrays don't have a length!

File: rllib/policy/sample_batch.py (hunk 10 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:181]: It's an np array that carries a scalar value. Sorry for the counter intuitive phrasing.

Screenshot 2022-11-09 at 10 31 28

File: rllib/policy/sample_batch.py (hunk 12 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:166]: Done!

File: rllib/policy/sample_batch.py (hunk 13 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:168]: Yep. Nones obviously don't have a length, but we are dropping them as you said.
But unsized arrays also don't have a length.
We currently support these as per our unit test. For example `SampleBatch({"obs": {"a":1, "b": 2}})` would fail otherwise.

Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:168]: Or `SampleBatch({"obs": {"a":np.array(1), "b": np.array(2)}})` for that matter

File: rllib/policy/sample_batch.py (hunk 11 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:158]: Which else are you referring to? Because the upper one belongs to the try/except clause, not the if.

This tree flattening also converts nested lists and therefore removes dimensions from them.
Screenshot 2022-11-09 at 10 51 18

File: rllib/policy/sample_batch.py (hunk 4 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py:164]: Screenshot 2022-11-09 at 11 05 27

I'm using startswith here

**Comment by ArturNiederfahrenhorst** [comment]
@gjoliver Since this operates on SampleBatch, which is a dict itself, I've simply added two lines to the unittest I had already included in this PR. I think that's a proper compromise between avoiding rewriting the testing logic and testing utility method and SampleBatch itself separately.

File: rllib/policy/sample_batch.py (hunk 13 above)
Comment (gjoliver) [reply rllib/policy/sample_batch.py]: oh, that i scalar np array. I see.

File: rllib/policy/sample_batch.py (hunk 11 above)
Comment (gjoliver) [reply rllib/policy/sample_batch.py]: oh, I was saying maybe ``v_list = tree.flatten(v)`` will just work here? do you really need ``[v]`` here?

**Review by gjoliver — APPROVED** [review]
can you quickly fix lint

!!! No API stability annotation found for:
--
  | ray.rllib.policy.sample_batch.attempt_count_timesteps

thanks for the thoughtful change.

File: rllib/policy/sample_batch.py (hunk 11 above)
Comment (ArturNiederfahrenhorst) [reply rllib/policy/sample_batch.py]: Ah right, changed! Thanks :)
